src/ast/passes/codegen_llvm.cpp

```diff
@@ -3807,6 +3820,49 @@ void CodegenLLVM::createIncDec(Unop &unop)
   }
 }
 
+Function *CodegenLLVM::createMapLenCallback()
+{
+  // The goal is to produce the following code:
+  //
+  // static int cb(struct map *map, void *key, void *value, void *ctx)
+  // {
+  //   int *len = (int *)ctx;
+  //   (*len)++;
+  //   return 0;
+  // }
+  auto saved_ip = b_.saveIP();
+
+  std::array<llvm::Type *, 4> args = {
+    b_.getInt8PtrTy(), b_.getInt8PtrTy(), b_.getInt8PtrTy(), b_.getInt8PtrTy()
+  };
+
+  FunctionType *callback_type = FunctionType::get(b_.getInt64Ty(), args, false);
+
+  Function *callback = Function::Create(callback_type,
+                                        Function::LinkageTypes::InternalLinkage,
+                                        "map_len_cb",
+                                        module_.get());
+
+  callback->setDSOLocal(true);
+  callback->setVisibility(llvm::GlobalValue::DefaultVisibility);
+  callback->setSection(".text");
+  debug_.createFunctionDebugInfo(*callback);
+
+  auto *bb = BasicBlock::Create(module_->getContext(), "", callback);
+  b_.SetInsertPoint(bb);
+
+  Value *ctx = callback->getArg(3);
+  auto len = b_.CreateBitCast(ctx, b_.getInt64Ty()->getPointerTo());
+  b_.CreateStore(
+      b_.CreateAdd(b_.CreateLoad(b_.getInt64Ty(), len), b_.getInt64(1)), len);
```

Comment:
This assumes that `len` is a uint64 right? Maybe more generic to get type from llvm? If not, perhaps add an `assert()` on that condition

Reply:
How would you want to do either of those? The callback receives `void *ctx` so I don't see many ways to check or assert the pointed type.

Reply:
Ah yeah you're right.